src/i2c/src/i2c.rs

```diff
@@ -0,0 +1,548 @@
+// I2C backend device
+//
+// Copyright 2021 Linaro Ltd. All Rights Reserved.
+//          Viresh Kumar <viresh.kumar@linaro.org>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use libc::{c_ulong, ioctl, EADDRINUSE, EADDRNOTAVAIL, EINVAL};
+use std::fs::{File, OpenOptions};
+use std::os::unix::io::AsRawFd;
+use vmm_sys_util::errno::{errno_result, Error, Result};
+
+// The type of the `req` parameter is different for the `musl` library. This will enable
+// successful build for other non-musl libraries.
+#[cfg(target_env = "musl")]
+type IoctlRequest = libc::c_int;
+#[cfg(not(target_env = "musl"))]
+type IoctlRequest = c_ulong;
+
+/// Linux I2C/SMBUS definitions
+/// IOCTL commands
+const I2C_SLAVE: IoctlRequest = 0x0703; // Use this slave address
```

Comment:
Does virito-i2c spec define any method to tune timeout/retries parameters?

Reply:
> Does virito-i2c spec define any method to tune timeout/retries parameters?

https://github.com/oasis-tcs/virtio-spec/blob/master/virtio-i2c.tex

Not really. But what exactly were you looking for ?

Reply:
> > Does virito-i2c spec define any method to tune timeout/retries parameters?
> 
> https://github.com/oasis-tcs/virtio-spec/blob/master/virtio-i2c.tex
> 
> Not really. But what exactly were you looking for ?

I'm think about the vhost-user service model. Do we have plan to support 1 vhost-user-i2c daemon for multiple virtual devices? Or we only support 1 daemon:1 device model?
If we have plan to support 1:N model, we should take care of the timeout for data rw IOCTLs, because I2C bus is slow and not so reliable.

Reply:
> I'm think about the vhost-user service model. Do we have plan to support 1 vhost-user-i2c daemon for multiple virtual devices? Or we only support 1 daemon:1 device model?
> If we have plan to support 1:N model, we should take care of the timeout for data rw IOCTLs, because I2C bus is slow and not so reliable.

Right now this supports one backend running at host, which supports multiple guests, with each guest containing a single virtual virtio I2C device. I am not sure why would we need to have multiple virtual I2C devices per guest.

Were you talking about this only ?

Reply:
Got it, the model is one vm one virtio-i2c device, one daemon for mutli virtio-i2c backend devices.
Then next question is one worker thread to serve one backend device or shared worker threads to serve all backend devices?
Due the nature of I2C/SMbus, the ioctl to read data from/write data to I2C controller may take a relatively long time. So will there be congestion caused by limited worker threads?

Reply:
> Got it, the model is one vm one virtio-i2c device, one daemon for mutli virtio-i2c backend devices.
> Then next question is one worker thread to serve one backend device or shared worker threads to serve all backend devices?
> Due the nature of I2C/SMbus, the ioctl to read data from/write data to I2C controller may take a relatively long time. So will there be congestion caused by limited worker threads?

Okay, I may have given a bit confusing answer earlier :(

So there is going to be only one instance of this binary running on the host machine.
But if you see main.rs, I am creating a separate Listener and VhostUserDaemon for every guest, each of which should be single threaded. So if there are 5 guests (i.e. 5 virtual i2c devices, 1 per guest), then this binary will start 5 different daemons, each of them should get a thread I believe..

Is that good enough ?

Reply:
Thanks for explanation, that makes sense:)